### PR TITLE
Deselect Row on Transaction Redo

### LINKED ISSUE #5673: When redo transactions the deleted row stays selected
## Description  
When redo transactions the deleted row stays selected

 * igniteui-angular version: 8.2
 * browser: n/a

## Steps to reproduce  

1. Open dev sample treeGrid
2. Select first row and click button 'Delete selected row'
3. Click undo button
4. Select same row
5. Click redo button

## Result  
The row is deleted but it is still selected

## Expected result  
The row should not be selected

## Attachments  
Attach a sample if available, and screenshots, if applicable.  
image

### PR DESCRIPTION
**Pull request by 3phase**
Closes #5673   

### Additional information (check all that apply):
 - [x] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/services/transaction/transaction.ts [hunk 1]
```diff
@@ -26,6 +35,11 @@ export interface State {
     type: TransactionType;
 }
 
+export interface IStateUpdateEvent {
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/services/transaction/transaction.ts:38]: we do not need `I` in front of the interface. In type script given class can `extend` other class or `implement` an interface. This is why when you read the class signature you know what class it extends and what interfaces it implements.

File: projects/igniteui-angular/src/lib/services/transaction/transaction.ts [hunk 2]
```diff
@@ -26,6 +35,11 @@ export interface State {
...
+    origin: TransactionEventOrigin;
+    actions?: any;
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/services/transaction/transaction.ts:40]: we can strongly type this. Please add interface `Action` like this:
```
export interface Action {
    transaction: Transaction;
    recordRef: any;
}
```
Note in `IgxTransactionService.redo` method there is `useInUndo` parameter in the action type - this is a leftover, and you can safely remove it.

File: projects/igniteui-angular/src/lib/services/transaction/igx-transaction.ts [hunk 3]
```diff
@@ -127,7 +127,7 @@ export class IgxTransactionService<T extends Transaction, S extends State> exten
             this._undoStack.push(actions);
             this._redoStack = [];
 
-            this.onStateUpdate.emit();
+            this.onStateUpdate.emit({ origin: TransactionEventOrigin.END});
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/services/transaction/igx-transaction.ts:130]: you can add the actions to the event arguments here too

File: projects/igniteui-angular/src/lib/services/transaction/igx-transaction.ts [hunk 4]
```diff
@@ -47,7 +47,7 @@ export class IgxTransactionService<T extends Transaction, S extends State> exten
         if (!this._isPending) {
             this._undoStack.push([{ transaction, recordRef }]);
             this._redoStack = [];
-            this.onStateUpdate.emit();
+            this.onStateUpdate.emit({ origin: TransactionEventOrigin.ADD });
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/services/transaction/igx-transaction.ts:50]: what we push in `_unodStack` is actually an `action`. Add this action to the event arguments.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,14 @@
 
 All notable changes for each version of this project will be documented in this file.
 
+## 10.0.0
+
+### General
+- `IgxTransaction` - The `onStateUpdate` now emits with information of its origin. The emitted value is of type `IStateUpdateEvent`, which has two properies:
+    - `origin` - a mandatory property. It can vary within the values of the `TransactionEventOrigin` interface; 
+    - `actions` - an optional property that contains information about the transactions, that caused the emission of the event. It is used by `undo` and `redo`.
```
Comment (wnvko) [inline CHANGELOG.md:10]: we will use the `actions` in every place where event emits - no need to clarify it is used in...

File: projects/igniteui-angular/src/lib/services/transaction/transaction.ts (hunk 2 above)
Comment (wnvko) [inline projects/igniteui-angular/src/lib/services/transaction/transaction.ts:40]: you can make this mandatory too and send empty array in `clear` method. For all other places you have the necessary actions.

**Comment by 3phase** [comment]
@wnvko thanks for your suggestions, I think I have resolved them. Could you take it a look again, please?

File: projects/igniteui-angular/src/lib/services/transaction/transaction.ts [hunk 6]
```diff
@@ -1,11 +1,20 @@
 import { EventEmitter } from '@angular/core';
+import { ThrowStmt } from '@angular/compiler';
 
 export enum TransactionType {
     ADD = 'add',
     DELETE = 'delete',
     UPDATE = 'update'
 }
 
+export enum TransactionEventOrigin {
```
Comment (skrustev) [inline projects/igniteui-angular/src/lib/services/transaction/transaction.ts:10]: @wnvko
Looking at the listed event origins for onStateUpdate I don't see when it is emitted from a commit. Looking at the IgxTransactionService I also didn't notice any place on commit that the onStateUpdate is called except on clear. Is this the expected behavior of the onStateUpdate event?

Also due to the IgxHierarchicalTransactionService commit calling clear() a second time when childDataKey is not provided, the onStateUpdate is emitted two times which I also am not sure if it's a valid scenario anyway in the igxTreeGrid but just asking.

Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/transaction/transaction.ts:10]: @skrustev 
Calling `onStateUpdate` in `clear` is OK from my point of view.
Calling `clear` in `IgxHierarchicalTransactionService.commit` is a bug. You can log this, as I do not thing it is OK to fix it in this feature PR.

Comment (skrustev) [reply projects/igniteui-angular/src/lib/services/transaction/transaction.ts:10]: @wnvko 
Ok so if the user wants to perform action on `commit` he will have to check for the `clear` event origin ? It's kinda weird to me but if you think it's ok I am fine as well.

Comment (3phase) [reply projects/igniteui-angular/src/lib/services/transaction/transaction.ts:10]: @skrustev @wnvko I have logged a separate issue (#7615) to address that problem.

**Pull request merged**